docs/testing-guidelines.md

```diff
@@ -0,0 +1,71 @@
+# Testing Guidelines for UDS Marketplace Apps
+
+## Introduction
+
+This document outlines the testing guidelines and best practices for UDS Marketplace Apps. These guidelines aim to ensure consistency, reliability, and quality across all UDS packages.
+
+## 1. Journey Testing
+
+Journey testing is crucial for ensuring that UDS packages function correctly within their deployment environment and integrate properly with other UDS components.
+
+#### **Definition**
+A **journey** in the context of UDS package testing is defined as:
+> The minimal and least fragile way to test all features of an application that could be broken from how we deploy it.
+
+#### **Key Principles**
+- Focus on testing the deployment, not the application itself (that's the upstream developer's job).
+- Ensure that our deployment method doesn't break key functionalities.
+- Test integration with UDS core components (e.g., Istio, Keycloak).
+
+#### **Implementation Guidelines**
+- Design tests that cover essential functionality without being overly complex.
+- Focus on deployment-related issues (e.g., network policies, SSO access, cluster resource access).
+- Handle limitations due to licensing or other constraints by documenting them and implementing what testing is possible.
+
+#### **Tools**
+- UI Testing: Playwright, Cypress
+- Non-UI Testing: Jest
+- Custom Scripts: Bash or other scripting languages as needed
+
+## 2. Upgrade Testing
+Ensure that the current development package works when deployed over the previously released version.
+
+#### **Key Considerations**
+- Test data migration and persistence.
+- Verify that configurations are properly carried over or updated.
+- Check for any breaking changes in APIs or integrations.
+
+## 3. Linting and Static Analysis
+
+Use appropriate tooling to ensure code quality and consistency.
+
+#### **Recommended Tools**
+- YAML Linting: `yamllint`
+- Zarf Package Linting: `zarf dev lint`
+
+## **Best Practices**
+
+1. **Consistency**: Use similar testing approaches across different UDS packages where possible.
+2. **Documentation**: Clearly document what each test is checking and why.
+3. **Maintenance**: Keep tests updated as the application and deployment methods evolve.
+4. **CI/CD Integration**: Incorporate all types of tests into your CI/CD pipeline.
+5. **Error Handling**: Ensure tests fail gracefully and provide clear error messages.
+6. **Idempotency**: Design tests to be repeatable and produce consistent results.
```

Comment:
I love idempotency but I'm not sure that's what we're capturing in the description of it.  Do we explicitly mean "repeated runs of the same test provide the same result" or do we mean "our tests are isolated and durable enough that regardless of ordering our tests get the same results"?

Reply:
I would say the latter, but I am also not super familiar with all of the testing strategies here yet. I am assuming that given the dynamic / async nature of some of these workloads that they might not ever be "repeatable" in the same order, but should be "repeatable" in terms of output. If that sounds good let me know and I can clean up the above!

Reply:
Should now be incorporated: https://github.com/defenseunicorns/uds-common/pull/205/commits/18853a49170b3cb54ba43d0f6f58d7fb6d9accf1